### ISSUE TITLE
Add Precision-Recall-Gain curve and Area Under Precision Recall Gain curve

### ISSUE THREAD

**Issue by siemdejong** [Lightning-AI/torchmetrics#1775]
## 🚀 Feature

Add Precision-Recall-Gain (PRG) curve as a new feature with the same interface as the Precision-Recall (PR) curve.

Along with PRG, the Area Under the Precision Recall Gain curve (AUPRG) can be calculated, like is done `AveragePrecision`.

### Motivation

The PR curve has some caveats as described in [1]. PRG aims to fix these problems:
1. baselines are non-universal
2. interpolation is non-linear
3. F-isometrics are non-linear
4. Pareto-front is non-convex
5. Area under PR curve does not relate to the expected F + there is an unachievable region

In particular, the area under the PR curve is demonstrated to sometimes favour models that result in lower F1-scores. The PRG curve will ultimately result in better model selection.

### Pitch

A Torchmetrics implementation of the PRG curve that has the same interface as the PR curve would aid in better model selection.

```python
>>> pred = torch.tensor([0, 0.1, 0.8, 0.4])
>>> target = torch.tensor([0, 1, 1, 0])
>>> prg_curve = PrecisionRecallGainCurve(task="binary")
>>> precision_gain, recall_gain, thresholds = prg_curve(pred, target)
>>> precision_gain
tensor([1.0000, 0.0000, 0.5000, 0.0000])
>>> recall_gain
tensor([0.0000,   0.0000,   1.0000,   1.0000])
>>> thresholds
...
```

Precision-Gain (PG) and Recall-Gain (RG) can be calculated as

$$
PG = 1 - \frac{tp + fn}{fp + tn} \cdot \frac{fp}{tp},
$$

and 

$$
RG = 1 - \frac{tp + fn}{fp + tn} \cdot \frac{fn}{tp}.
$$

AUPRG can be calculated as done with `AveragePrecision`, but only accounting for the area in PR & RG $\in [0, 1]$.

It would be even more awesome if PRG can be extended to the multiclass/multilabel case.

### Alternatives

The original authors of [1] have developed a package, [pyprg](https://github.com/meeliskull/prg) (which is out-of-date with dependencies). 
```
pip instal pyprg
```
Then,
```python
from prg import prg
prg_curve = prg.create_prg_curve(labels=targets, scores=prediction)
precision_gain = prg_curve["precision_gain"]
recall_gain = prg_curve["recall_gain"]
auprg= prg.calc_auprg(prg_curve)
```

### Additional context

[1] Flach & Kull. http://people.cs.bris.ac.uk/~flach/PRGcurves/PRcurves.pdf

**Comment by SkafteNicki** [comment]
Hi @siemdejong, thanks for raising this issue.
A couple of questions maybe:
* How commonly is this metric used? I have not heard or seen it before in any papers I read.
* It is good that there is a package to compare against if we make our own implementation, but then I see issues like https://github.com/meeliskull/prg/issues/7 and wonder how stable the implementation is?

**Comment by siemdejong** [comment]
- The metric is not (yet) commonly used. Obvious reasons might be that 1) people simply do not know about it, 2) it takes an extra step to calculate the plot, 3) no good implementation is available.
- I have not tested the implementation thoroughly, so I cannot make arguments on the stability of the official implementation.

For another writeup about gain metrics, see https://snorkel.ai/improving-upon-precision-recall-and-f1-with-gain-metrics/

Maybe an interesting discussion on scikit-learn and gain metrics: https://github.com/scikit-learn/scikit-learn/pull/24121

**Comment by arijitde92** [comment]
Hi, can I contribute in this issue?

**Comment by SkafteNicki** [comment]
Hi @arijitde92,
Feel free to make a contribution on this topic :)

**Comment by e-pet** [comment]
It might be of interest that I implemented these metrics (basically from scratch) in https://github.com/FraunhoferMEVIS/meval/tree/main, specifically here: https://github.com/FraunhoferMEVIS/meval/blob/6b166e9ec5b893a6f3b29fb82ffb679fe9d0b0fc/meval/metrics/_metrics.py#L290
